samples: Bluetooth: Broadcast audio source USB headset on nRF52 Series

cvinayak opened the pull request:
Broadcast audio source USB headset on nRF52 Series with USB support.

cvinayak committed:
samples: Bluetooth: Broadcast audio source USB headset on nRF52 Series

Broadcast audio source USB headset on nRF52 Series with USB
support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

larsgk commented:
This is awesome 😎 I would like to try it with an nRF52840 dongle too (if it works, you can transform any modern phone to be a broadcaster ;))

larsgk commented:
@cvinayak  What is missing to move this PR out of draft?

cvinayak commented:
> @cvinayak What is missing to move this PR out of draft?

i will need to test on a nRF52840 dongle, remove the nrf52833, have only overlay for nrf52840dongle (which should be sufficient), and update documentation. Hope to do these by end of this week, once In get a dongle to test.

cvinayak commented:
> > > @cvinayak What is missing to move this PR out of draft?
> > 
> > 
> > i will need to test on a nRF52840 dongle, remove the nrf52833, have only overlay for nrf52840dongle (which should be sufficient), and update documentation. Hope to do these by end of this week, once I get a dongle to test.
> 
> also, for encrypted broadcast, I think only 1-2 (newest) nRF52x SoCs + nRF5340 work. For unencrypted, the dongle will be perfect. Do you have a dongle to test with?

Yes, only newer nRF5 SoCs will truely support encrypted Broadcast Audio, but any nRF52x and nRF53x series will work for cleartext broadcast audio.

I have now tested on nrf52840dongle_nrf52840  boards, please build your self and revert back if any issues...

larsgk commented:
@cvinayak thanks for testing on the dongle - how was the quality?

Btw .. there is a PR with a bump of liblc3 https://github.com/zephyrproject-rtos/zephyr/pull/66017

Would it be possible for you to do a quick smoke test on the DK with the new version?

cvinayak commented:
> @cvinayak thanks for testing on the dongle - how was the quality?
> 
> Btw .. there is a PR with a bump of liblc3 #66017
> 
> Would it be possible for you to do a quick smoke test on the DK with the new version?

Did a quick test, no regressions.

larsgk commented:
> > @cvinayak thanks for testing on the dongle - how was the quality?
> > Btw .. there is a PR with a bump of liblc3 #66017
> > Would it be possible for you to do a quick smoke test on the DK with the new version?
> 
> Did a quick test, no regressions.

and I just flashed a dongle.  Clear audio reception in the other end... this... is awesome!!! Thanks @cvinayak

Thalley commented on samples/bluetooth/broadcast_audio_source/boards/nrf52833dk_nrf52833.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_USE_USB_AUDIO_INPUT=y
+CONFIG_USB_DEVICE_PRODUCT="Zephyr Broadcast Source"
+
+# Two streams in one subgroup (stereo)
+CONFIG_BT_BAP_BROADCAST_SRC_SUBGROUP_COUNT=1
```
We could consider making this the default instead

Thalley commented on samples/bluetooth/broadcast_audio_source/boards/nrf52833dk_nrf52833.conf:
```diff
@@ -0,0 +1,8 @@
+
+# Use USB Audio as input
+CONFIG_USE_USB_AUDIO_INPUT=y
+CONFIG_USB_DEVICE_PRODUCT="Zephyr Broadcast Source"
```
We could consider making this the default

cvinayak replied:
Not all SoCs support USB, lets keep the USB Kconfig in board configs for now.

Thalley replied:
Fair

Thalley commented on samples/bluetooth/broadcast_audio_source/boards/nrf52833dk_nrf52833.overlay:
```diff
@@ -0,0 +1,15 @@
+	compatible = "nordic,nrf-usbd";
+	status = "okay";
+
+	hs_0: hs_0 {
+		compatible = "usb-audio-hs";
+		mic-feature-mute;
+		mic-channel-l;
+		mic-channel-r;
+
+		hp-feature-mute;
+		hp-channel-l;
+		hp-channel-r;
+	};
```
These exists in the `.dtsi` for the nRF 5340 ADK;

Why is it different for this board?

larsgk replied:
My guess is that the Audio DK is considered to use USB for audio more than the standard DK.

cvinayak replied:
only CDC-ACM included by default in nRF SoC's with USB support, hence need an explicit overlay in this sample.

Thalley commented on samples/bluetooth/broadcast_audio_source/prj.conf:
```diff
@@ -13,3 +13,10 @@ CONFIG_BT_ISO_TX_BUF_COUNT=4
 
 CONFIG_BT_DEVICE_NAME="Broadcast Audio Source"
+
+# program stack and CPU usage
+CONFIG_THREAD_ANALYZER=y
+CONFIG_THREAD_ANALYZER_USE_PRINTK=y
+CONFIG_THREAD_ANALYZER_AUTO=y
+CONFIG_THREAD_ANALYZER_AUTO_INTERVAL=5
+CONFIG_THREAD_NAME=y
```
This should not be enabled by default in the sample IMO

cvinayak replied:
Removed

cvinayak committed:
samples: Bluetooth: Update buffer counts for Broadcast Audio Source

Updaye buffer counts and semaphore initial value so that the
Broadcast Audio Source can utilize any use of broadcast
pre-transmissions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Updated to take advantage of per-transmissions.

Thalley commented on samples/bluetooth/broadcast_audio_source/prj.conf:
```diff
@@ -7,9 +7,10 @@ CONFIG_BT_BAP_BROADCAST_SOURCE=y
+# Two streams in one subgroup (stereo)
+CONFIG_BT_BAP_BROADCAST_SRC_SUBGROUP_COUNT=1
 
+CONFIG_BT_ISO_TX_BUF_COUNT=6
```
This effectively gives us 3 bufs per stream; wasn't 2 enough?

cvinayak replied:
2 buffers will be held per ISO interval and hence 1 additional for the next ISO interval.

cvinayak replied:
Tx buffers live/allocated across ISO Interval plus the delays; an SDU's PDU Tx buffer can remain allocated until all the required ISO intervals are exhausted.

In the case of 1 pre-transmission count and pre-transmission offset of 1, the SDU's PDU Tx buffer remains allocated for 2 adjacent ISO interval plus delay and a latency upto one ISO interval. Hence, to have new SDUs ready for subsequent ISO interval, additional n free PDU Tx buffers need be available for the application to be able to enqueue in time for the transmission.
